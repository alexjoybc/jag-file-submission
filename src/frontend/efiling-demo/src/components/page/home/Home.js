import React, { useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Header, Footer, Input, Button } from "shared-components";
import { FilePond, registerPlugin } from "react-filepond";

// Import FilePond styles
import "filepond/dist/filepond.min.css";

// Import the Image EXIF Orientation and Image Preview plugins
import FilePondPluginImageExifOrientation from "filepond-plugin-image-exif-orientation";
import FilePondPluginImagePreview from "filepond-plugin-image-preview";
import "filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css";

import { propTypes } from "../../../types/propTypes";
import "../page.css";

// Register the plugins
registerPlugin(FilePondPluginImageExifOrientation, FilePondPluginImagePreview);

const urlBody = {
  clientApplication: {
    displayName: "Demo App",
    type: "app"
  },
  filingPackage: {
    court: {
      location: "string",
      level: "string",
      class: "string",
      division: "string",
      fileNumber: "string",
      participatingClass: "string"
    },
    documents: [
      {
        name: "string",
        description: "string",
        type: "string"
      }
    ]
  },
  navigation: {
    success: {
      url: `${window.location.origin}/efiling-demo/success`
    },
    error: {
      url: `${window.location.origin}/efiling-demo/error`
    },
    cancel: {
      url: `${window.location.origin}/efiling-demo/cancel`
    }
  }
};

const input = {
  label: "Account GUID",
  id: "textInputId",
  styling: "editable-white",
  isRequired: true,
  placeholder: "77da92db-0791-491e-8c58-1a969e67d2fa"
};

const generatePackageData = files => {
  const formData = new FormData();
  const documentData = [];

  for (let i = 0; i < files.length; i += 1) {
    formData.append("files", files[i].file);
    documentData.push({
      name: files[i].file.name,
      description: "file description",
      type: "file type"
    });
  }

  const updatedUrlBody = {
    ...urlBody,
    filingPackage: {
      ...urlBody.filingPackage,
      documents: documentData
    }
  };

  return { formData, updatedUrlBody };
};

export const eFilePackage = (files, accountGuid, setErrorExists) => {
  const { formData, updatedUrlBody } = generatePackageData(files);

  axios
    .post("/submission/documents", formData, {
      headers: {
        "X-Auth-UserId": accountGuid,
        "Content-Type": "multipart/form-data"
      }
    })
    .then(response => {
      const { submissionId } = response.data;

      axios
        .post(`/submission/${submissionId}/generateUrl`, updatedUrlBody, {
          headers: { "X-Auth-UserId": accountGuid }
        })
        .then(({ data: { efilingUrl } }) => {
          window.open(efilingUrl, "_self");
        })
        .catch(() => {
          setErrorExists(true);
        });
    })
    .catch(() => {
      setErrorExists(true);
    });
};

export default function Home({ page: { header } }) {
  const [errorExists, setErrorExists] = useState(false);
  const [accountGuid, setAccountGuid] = useState(null);
  const [files, setFiles] = useState([]);

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="content col-md-12">
          <Input input={input} onChange={setAccountGuid} />
          <br />
          <br />
          <FilePond
            files={files}
            allowMultiple
            onupdatefiles={setFiles}
            labelIdle='Drag and Drop your files or <span class="filepond--label-action">Browse</span>'
          />
          <br />
          <Button
            onClick={() => eFilePackage(files, accountGuid, setErrorExists)}
            label="E-File my Package"
            styling="normal-blue btn"
            testId="generate-url-btn"
          />
          <br />
          {errorExists && (
            <p className="error">
              An error occurred while generating the URL. Please try again.
            </p>
          )}
        </div>
      </div>
      <Footer />
    </main>
  );
}

Home.propTypes = {
  page: PropTypes.shape({
    header: propTypes.header
  }).isRequired
};
