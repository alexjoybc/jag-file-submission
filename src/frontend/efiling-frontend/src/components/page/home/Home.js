import React, { useState } from "react";
import PropTypes, { element } from "prop-types";
import queryString from "query-string";
import { useLocation } from "react-router-dom";
import axios from "axios";
import Header, { Footer, Loader } from "shared-components";
import CSOStatus from "../../composite/csostatus/CSOStatus";

import "../page.css";

export const saveNavigationToSession = ({ cancel, success, error }) => {
  if (cancel.url) sessionStorage.setItem("cancelUrl", cancel.url);
  if (success.url) sessionStorage.setItem("successUrl", success.url);
  if (error.url) sessionStorage.setItem("errorUrl", error.url);
};

// make call to submission/{id} to get the user and navigation details
const checkCSOAccountStatus = (
  submissionId,
  setCsoAccountExists,
  setShowLoader
) => {
  axios
    .get(`/submission/${submissionId}`)
    .then(({ data: { userDetails, navigation } }) => {
      saveNavigationToSession(navigation);

      if (userDetails.accounts) {
        const csoAccountExists = userDetails.accounts.some(
          element => element.type === "CSO"
        );
        if (csoAccountExists) setCsoAccountExists(true);
      }
      setShowLoader(false);
    })
    .catch(() => {});
};

export default function Home({ page: { header } }) {
  const [showLoader, setShowLoader] = useState(true);
  const [csoAccountExists, setCsoAccountExists] = useState(false);
  const location = useLocation();
  const queryParams = queryString.parse(location.search);

  checkCSOAccountStatus(
    queryParams.submissionId,
    setCsoAccountExists,
    setShowLoader
  );

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="content col-md-10">
          {showLoader && <Loader page />}
          {!showLoader && <CSOStatus accountExists={csoAccountExists} />}
        </div>
      </div>
      <Footer />
    </main>
  );
}

Home.propTypes = {
  page: PropTypes.shape({
    header: PropTypes.shape({
      name: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
};
