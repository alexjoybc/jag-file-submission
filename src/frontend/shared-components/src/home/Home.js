import React from "react";

export default function Home() {
  return (
    <div>
      <p>Shared Components</p>
    </div>
  );
}
