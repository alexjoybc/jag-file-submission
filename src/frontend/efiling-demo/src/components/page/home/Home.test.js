import React from "react";
import { createMemoryHistory } from "history";
import axios from "axios";
import {
  render,
  waitFor,
  getByText,
  fireEvent,
  getAllByRole,
} from "@testing-library/react";
import Home, { eFilePackage } from "./Home";

const MockAdapter = require("axios-mock-adapter");

window.open = jest.fn();

const header = {
  name: "eFiling Demo Client",
  history: createMemoryHistory(),
};

const page = { header };

describe("Home", () => {
  let mock;
  const setErrorExists = jest.fn();
  const files = [
    {
      file: {
        name: "filename",
        type: "filetype",
      },
    },
  ];
  const submissionId = "123";
  const filingPackage = {
    documents: [files[0].file],
  };
  const token = "validJWT";

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  test("Component matches the snapshot", () => {
    const { asFragment } = render(<Home page={page} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("eFilePackage function displays an error message on page on failure of generateUrl call", async () => {
    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock.onPost("/submission/generateUrl").reply(400);

    eFilePackage(token, files, setErrorExists, filingPackage);

    await waitFor(() => {});

    expect(setErrorExists).toHaveBeenCalledWith(true);
  });

  test("eFilePackage function displays an error message on page on failure of uploadDocuments call", async () => {
    mock.onPost("/submission/documents").reply(400);

    eFilePackage(token, files, setErrorExists, filingPackage);

    await waitFor(() => {});

    expect(setErrorExists).toHaveBeenCalledWith(true);
  });

  test("eFilePackage function generates the proper documentData for the updated url body and redirects to frontend app on success", async () => {
    const efilingUrl = "example.com";

    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock
      .onPost(`/submission/${submissionId}/generateUrl`)
      .reply(200, { efilingUrl });

    eFilePackage(token, files, setErrorExists, filingPackage);

    await waitFor(() => {});

    expect(window.open).toHaveBeenCalledTimes(1);
  });

  test("eFilePackage functions returns error when no files uploaded", async () => {
    const { container } = render(<Home page={page} />);

    const textbox = getAllByRole(container, "textbox");

    fireEvent.change(textbox[0], {
      target: { value: JSON.stringify(filingPackage) },
    });

    fireEvent.click(getByText(container, "E-File my Package"));

    await waitFor(() => {});

    expect(
      getByText(
        container,
        "An error occurred while eFiling your package. Please make sure you upload at least one file and try again."
      )
    ).toBeInTheDocument();
    expect(setErrorExists).toHaveBeenCalledWith(true);
  });

  test("eFilePackage does not make axios call when no formdata present (due to incorrect filingPackage data)", () => {
    const improperFilingPackage = {
      documents: [{ name: "wrongname", type: "type" }],
    };

    eFilePackage(token, files, setErrorExists, improperFilingPackage);

    expect(setErrorExists).toHaveBeenCalledWith(true);
  });
});
