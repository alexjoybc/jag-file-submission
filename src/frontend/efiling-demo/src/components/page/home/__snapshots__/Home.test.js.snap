// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component matches the snapshot 1`] = `
<DocumentFragment>
  <main>
    <header>
      <nav
        class="container-fluid navbar navbar-expand-lg navbar-dark"
      >
        <div
          aria-labelledby="title"
          class="navbar-brand pointer"
          role="button"
          tabindex="0"
        />
        <img
          alt="B.C. Government Logo"
          class="img-fluid d-none d-md-block"
          height="44"
          src="[object Object]"
          width="181"
        />
        <img
          alt="B.C. Government Logo"
          class="img-fluid d-md-none"
          height="44"
          src="[object Object]"
          width="64"
        />
        <div
          aria-labelledby="title"
          class="pointer navbar-brand nav-item nav-link"
          role="button"
          tabindex="0"
        />
        <div
          class="navbar-brand"
          id="title"
        >
          eFiling Demo Client
        </div>
      </nav>
    </header>
    <div
      class="page"
    >
      <div
        class="content col-md-10"
      >
        <div>
          <label
            class="label"
            for="textInputId"
          >
            <div
              class="textinput_label"
            >
              Account GUID
            </div>
            <span
              class="mandatory"
              id="asterisk"
            >
              *
            </span>
            Â 
            <span
              class="note"
            />
          </label>
          <input
            class="editable_white"
            id="textInputId"
            placeholder="77da92db-0791-491e-8c58-1a969e67d2fa"
            type="text"
            value=""
          />
          <br />
          <span
            class="error"
          />
        </div>
        <br />
        <div>
          <button
            class="normal-blue btn"
            type="button"
          >
            Generate URL
          </button>
        </div>
        <br />
      </div>
    </div>
    <footer
      class="footer"
    >
      <nav
        class="navbar navbar-expand-sm navbar-dark justify-content-end"
      >
        <button
          class="navbar-toggler"
          data-target="#footerBar"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse flex-grow-0"
          id="footerBar"
        >
          <ul
            class="navbar-nav text-right"
          >
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca"
                rel="noopener noreferrer"
                target="_blank"
              >
                BC Government
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
                rel="noopener noreferrer"
                target="_blank"
              >
                Disclaimer
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/privacy"
                rel="noopener noreferrer"
                target="_blank"
              >
                Privacy
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/accessibility"
                rel="noopener noreferrer"
                target="_blank"
              >
                Accessibility
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/copyright"
                rel="noopener noreferrer"
                target="_blank"
              >
                Copyright
              </a>
            </li>
          </ul>
        </div>
      </nav>
    </footer>
  </main>
</DocumentFragment>
`;
